src/Microsoft.Framework.FileSystemGlobbing/Internal/MatcherContext.cs

```diff
@@ -74,7 +96,44 @@ private void Match(DirectoryInfoBase directory, string parentRelativePath)
             }
 
             // Request all the including and excluding patterns to pop their status stack.
-            PopFrame();
+            PopDirectory();
+        }
+
+        private void Predict()
+        {
+            _declaredLiteralFileSegments.Clear();
+            _declaredLiteralFolderSegments.Clear();
+            _declaredParentPathSegment = false;
+            _declaredWildcardPathSegment = false;
+
+            foreach (var include in _includePatternContexts)
+            {
+                include.Predict(DeclareInclude);
+            }
+        }
+
+        private void DeclareInclude(IPathSegment patternSegment, bool isLastSegment)
+        {
+            var literalSegment = patternSegment as LiteralPathSegment;
+            if (literalSegment != null)
+            {
+                if (isLastSegment)
+                {
+                    _declaredLiteralFileSegments.Add(literalSegment);
+                }
+                else
+                {
+                    _declaredLiteralFolderSegments.Add(literalSegment);
+                }
+            }
+            else if (patternSegment is ParentPathSegment)
+            {
+                _declaredParentPathSegment = true;
+            }
+            else if (patternSegment is WildcardPathSegment)
+            {
+                _declaredWildcardPathSegment = true;
+            }
         }
 
         private string CombinePath(string left, string right)
```

Comment:
We want this package to be self-contained because the `Microsoft.Framework.Runtime` will depend on it.